platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/workspace/impl/CacheBasedImmutableWorkspaceProvider.java

```diff
@@ -133,12 +104,24 @@ private static CleanupAction createCleanupAction(FileAccessTimeJournal fileAcces
     }
 
     @Override
-    public <T> T withWorkspace(String path, WorkspaceAction<T> action) {
-        return cache.withFileLock(() -> {
-            File workspace = new File(baseDirectory, path);
-            fileAccessTracker.markAccessed(workspace);
-            return action.executeInWorkspace(workspace, executionHistoryStore);
-        });
+    public ImmutableWorkspace getWorkspace(String path) {
+        File immutableWorkspace = new File(baseDirectory, path);
+        fileAccessTracker.markAccessed(immutableWorkspace);
+        return new ImmutableWorkspace() {
+            @Override
+            public File getImmutableLocation() {
+                return immutableWorkspace;
+            }
+
+            @Override
+            public <T> T withTemporaryWorkspace(TemporaryWorkspaceAction<T> action) {
+                // TODO Use Files.createTemporaryDirectory() instead
+                String temporaryLocation = path + "-" + UUID.randomUUID();
+                File temporaryWorkspace = new File(baseDirectory, temporaryLocation);
+                // TODO Do we need to track the temporary workspace in the access tracker?
```

Comment:
❔ Do we have an issue for that?

Reply:
Well spotted. We shouldn't need to, as if the access time is not recorded in the tracker, it will fall back to the file's modification time. I removed the TODO.